internal/expr/parser.go

```diff
@@ -265,28 +272,174 @@ func (p *Parser) parseIdentifier() (string, bool) {
+			if idCol, ok := p.parseIdentifierAsterisk(); ok {
+				return fullName{prefix: id, name: idCol}, true, nil
+			}
+		} else {
+			// A column name specified without a table prefix should be in name.
+			return fullName{name: id}, true, nil
+		}
+	}
+
+	cp.restore()
+	return fullName{}, false, nil
+}
+
+func (p *Parser) parseTarget() (fullName, bool, error) {
+	cp := p.save()
+
+	if p.skipByte('&') {
+		return p.parseGoFullName()
+	}
+
+	cp.restore()
+	return fullName{}, false, nil
+}
+
 // parseGoFullName parses a Go type name qualified by a tag name (or asterisk)
-// of the form "TypeName.col_name". On success it returns the parsed fullName,
+// of the form "&TypeName.col_name". On success it returns the parsed FullName,
 // true and nil. If a Go full name is found, but not formatted correctly, false
 // and an error are returned. Otherwise the error is nil.
 func (p *Parser) parseGoFullName() (fullName, bool, error) {
 	cp := p.save()
+
 	if id, ok := p.parseIdentifier(); ok {
-		if p.skipByte('.') {
-			if idField, ok := p.parseIdentifierAsterisk(); ok {
-				return fullName{id, idField}, true, nil
-			}
-			return fullName{}, false,
-				fmt.Errorf("invalid identifier near char %d", p.pos)
+		if !p.skipByte('.') {
+			return fullName{}, false, fmt.Errorf("go object near char %d not qualified", p.pos)
+		}
+
+		idField, ok := p.parseIdentifierAsterisk()
+		if !ok {
+			return fullName{}, false, fmt.Errorf("invalid identifier near char %d", p.pos)
 		}
-		return fullName{}, false,
-			fmt.Errorf("go object near char %d not qualified", p.pos)
+		return fullName{id, idField}, true, nil
 	}
+
 	cp.restore()
 	return fullName{}, false, nil
 }
 
-// parseInputExpression parses an input expression of the form $Type.name.
+// parseList takes a parsing function that returns a fullName and parses a
+// bracketed, comma seperated, list.
+func (p *Parser) parseList(parseFn func(p *Parser) (fullName, bool, error)) ([]fullName, bool, error) {
+	cp := p.save()
+
+	if !p.skipByte('(') {
+		return nil, false, nil
+	}
+
+	parenPos := p.pos
+	p.skipSpaces()
+
+	nextItem := true
+	var objs []fullName
+	for i := 0; nextItem; i++ {
+		if obj, ok, err := parseFn(p); ok {
+			objs = append(objs, obj)
+		} else if err != nil {
+			return nil, false, err
+		} else if i == 0 {
+			// If the first item is not what we are looking for, we exit.
+			cp.restore()
+			return nil, false, nil
+		} else {
+			// On subsequent items we return an error.
+			return nil, false, fmt.Errorf("column %d: invalid expression", p.pos)
+		}
+
+		p.skipSpaces()
+		if p.skipByte(')') {
+			return objs, true, nil
+		}
+
+		nextItem = p.skipByte(',')
+		p.skipSpaces()
+	}
+	return nil, false, fmt.Errorf("column %d: missing closing parentheses", parenPos)
+}
+
+// parseColumns parses a list of columns. For lists of more than one column the
+// columns must be enclosed in brackets e.g. "(col1, col2) AS &Person.*".
+func (p *Parser) parseColumns() ([]fullName, bool) {
+	cp := p.save()
```

Comment:
I believe we don't need to save/restore here either, because the entire functionality is implemented inside the two functions below, and only one of them can possibly work.

Reply:
Changed.